src/number/complete.rs

```diff
@@ -1357,29 +1358,35 @@ where
 ///
 /// assert_eq!(parser(&b"01AE"[..]), Ok((&b""[..], 0x01AE)));
 /// assert_eq!(parser(&b"abc"[..]), Ok((&b""[..], 0x0ABC)));
-/// assert_eq!(parser(&b"ggg"[..]), Err(Err::Error((&b"ggg"[..], ErrorKind::IsA))));
+/// assert_eq!(parser(&b"ggg"[..]), Err(Err::Error((&b"ggg"[..], ErrorKind::HexDigit))));
 /// ```
 #[inline]
-pub fn hex_u32<'a, E: ParseError<&'a [u8]>>(input: &'a [u8]) -> IResult<&'a [u8], u32, E> {
-  let (i, o) = crate::bytes::complete::is_a(&b"0123456789abcdefABCDEF"[..])(input)?;
+pub fn hex_u32<'a, I, E: ParseError<I>>(input: I) -> IResult<I, u32, E>
+where
+  I: AsBytes,
+  I: Clone,
+  I: InputIter,
+  I: InputLength,
+  I: Slice<RangeFrom<usize>> + Slice<RangeTo<usize>>,
+  <I as InputIter>::Item: AsChar + Copy,
+{
   // Do not parse more than 8 characters for a u32
-  let (parsed, remaining) = if o.len() <= 8 {
-    (o, i)
-  } else {
-    (&input[..8], &input[8..])
-  };
-
-  let res = parsed
-    .iter()
-    .rev()
-    .enumerate()
-    .map(|(k, &v)| {
-      let digit = v as char;
-      digit.to_digit(16).unwrap_or(0) << (k * 4)
-    })
-    .sum();
-
-  Ok((remaining, res))
+  fold_many_m_n(
+    1,
+    8,
+    |input: I| {
+      let c = match input.iter_elements().next() {
+        None => return Err(Err::Error(E::from_error_kind(input, ErrorKind::Eof))),
+        Some(c) => c,
+      };
+      match c.as_char().to_digit(16) {
+        Some(digit) => Ok((input.slice(c.len()..), digit)),
+        None => Err(Err::Error(E::from_error_kind(input, ErrorKind::HexDigit))),
+      }
+    },
+    || 0,
+    |acc, digit| (16 * acc) + digit,
+  )(input)
```

Comment:
I can rewrite this without any call to `fold_many_m_n` by using `input.iter_elements()` and manually enumerating and counting how many hex digits I find on the way, but I thought this was more elegant.

It might make sense to factor out the parsing lambda into its own method (kinda like hex_digit0, but guaranteed to take at most one character).

Reply:
I actually gave this rewrite a shot and IMO it looks nicer. Going to update with that.